spec/integration/integration_spec.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+require "spec_helper"
+
+RSpec.describe "multiple workers" do
+  def with_workers(n)
+    workers = Array.new(n) { Que::Worker.new }
+    worker_threads = workers.map { |worker| Thread.new { worker.work_loop } }
+
+    yield
+
+    workers.each(&:stop!)
+    worker_threads.each do |thread|
+      raise "timed out waiting for worker to finish!" unless thread.join(5)
+    end
+  end
+
+  # Wait for a maximum of [timeout] seconds for all jobs to be worked
+  def wait_for_jobs_to_be_worked(timeout: 10)
+    start = Time.now
+    loop do
+      break if QueJob.count == 0 || Time.now - start > timeout
+      sleep 0.1
+    end
+  end
+
+  context "with one worker and many jobs" do
+    it "works each job exactly once" do
```

Comment:
We use the phrase "exactly once" a few times in these tests. But from reading them, I don't think that we have a  guarantee that the jobs have been worked exactly once. E.g, we enqueue 1 job, spin up 5 workers and assert that we have no more jobs in the queue. But we do not assert that only one worker actually ran the job, not sure how we test this/if we can test this at all?

Reply:
I have amended this so we verify the number of elements added to the job log, which should indicate how many times each job was run. The tests now fail if you've double enqueued any of the jobs.

Nice spot!